exporter/opentelemetry-exporter-otlp-json-common/src/opentelemetry/exporter/otlp/json/common/_internal/trace_encoder/__init__.py

```diff
@@ -0,0 +1,348 @@
+
+        if instrumentation_scope_hashcode not in scope_spans:
+            scope_spans[instrumentation_scope_hashcode] = {
+                "scope": _encode_instrumentation_scope(
+                    span.instrumentation_scope
+                ),
+                "spans": [],
+                "schemaUrl": (
+                    span.instrumentation_scope.schema_url
+                    if hasattr(span.instrumentation_scope, "schema_url")
+                    else ""
+                ),
+            }
+
+        scope_spans[instrumentation_scope_hashcode]["spans"].append(
+            _encode_span(span)
+        )
+
+    # Transform resource_spans dict to list for proper JSON output
+    resource_spans_list = []
+    for resource_span_data in resource_spans.values():
+        scope_spans_list = []
+        for scope_span_data in resource_span_data["scopeSpans"].values():
+            scope_spans_list.append(scope_span_data)
+
+        resource_span_data["scopeSpans"] = scope_spans_list
+        resource_spans_list.append(resource_span_data)
+
+    return {"resourceSpans": resource_spans_list}
+
+
+def _compute_resource_hashcode(resource: Resource) -> str:
+    """Computes a hashcode for the resource based on its attributes."""
+    if not resource.attributes:
+        return ""
+    # Simple implementation: use string representation of sorted attributes
+    return str(sorted(resource.attributes.items()))
+
+
+def _compute_instrumentation_scope_hashcode(
+    scope: InstrumentationScope,
+) -> str:
+    """Computes a hashcode for the instrumentation scope."""
+    if scope is None:
+        return ""
+    return f"{scope.name}|{scope.version}"
+
+
+def _encode_resource(resource: Resource) -> Dict[str, Any]:
+    """Encodes a resource into OTLP JSON format."""
+    if not resource:
+        return {"attributes": []}
+
+    return {
+        "attributes": _encode_attributes(resource.attributes),
+        "droppedAttributesCount": 0,  # Not tracking dropped attributes yet
+    }
+
+
+def _encode_instrumentation_scope(
+    scope: Optional[InstrumentationScope],
+) -> Dict[str, Any]:
+    """Encodes an instrumentation scope into OTLP JSON format."""
+    if scope is None:
+        return {"name": "", "version": ""}
+
+    return {
+        "name": scope.name or "",
+        "version": scope.version or "",
+        "attributes": [],  # Not using attributes for scope yet
+        "droppedAttributesCount": 0,
+    }
+
+
+def _encode_span(span: ReadableSpan) -> Dict[str, Any]:
+    """Encodes a span into OTLP JSON format."""
+    # Convert trace_id and span_id to base64
+    trace_id_bytes = span.context.trace_id.to_bytes(16, "big")
+    span_id_bytes = span.context.span_id.to_bytes(8, "big")
+
+    parent_id = ""
+    # Handle different span implementations that might not have parent_span_id
+    if hasattr(span, "parent_span_id") and span.parent_span_id:
+        parent_id = base64.b64encode(
```

Comment:
hex, not base64

https://opentelemetry.io/docs/specs/otlp/#json-protobuf-encoding